### PR TITLE
[MAINT] remove obsolete version comparison

### LINKED ISSUE #3706: Check if special handling of older matplotlib are still needed
Maybe leave for another PR but maybe this patch and hence test are not needed anymore given that we support mpl>=3.3.0

_Originally posted by @ymzayek in https://github.com/nilearn/nilearn/pull/3648#discussion_r1165466003_

### LINKED ISSUE #3793: check where `_compare_version` is used in codebase referencing package versions we don't support anymore and remove extraneous code
Note to self: check where `_compare_version` is used in codebase referencing package versions we don't support anymore and remove extraneous code

_Originally posted by @ymzayek in https://github.com/nilearn/nilearn/pull/3790#discussion_r1246640452_

List of occurrence of compare_version to potentially check:

```bash
$ find nilearn -name '*.py' -exec grep -H "_compare_version" {} \;

nilearn/image/resampling.py:from nilearn._utils import _compare_version
nilearn/image/resampling.py:        if _compare_version(scipy.__version__, ">=", "0.18"):
nilearn/image/resampling.py:            if _compare_version(scipy.__version__, ">=", "0.18"):
nilearn/decoding/tests/test_decoder.py:from nilearn._utils import _compare_version
nilearn/decoding/tests/test_decoder.py:    if _compare_version(sklearn.__version__, ">", "1.2.2"):
nilearn/_utils/__init__.py:    _compare_version,
nilearn/_utils/__init__.py:    "_compare_version",
nilearn/_utils/tests/test_helpers.py:def test_compare_version(version_a, operator, version_b):
nilearn/_utils/tests/test_helpers.py:    assert helpers._compare_version(version_a, operator, version_b)
nilearn/_utils/tests/test_helpers.py:def test_compare_version_error():
nilearn/_utils/tests/test_helpers.py:        match="'_compare_version' received an unexpected operator <>.",
nilearn/_utils/tests/test_helpers.py:        helpers._compare_version("0.1.0", "<>", "1.1.0")
nilearn/_utils/testing.py:from nilearn._utils import _compare_version
nilearn/_utils/testing.py:        if _compare_version(sklearn.__version__, "<", "0.22"):
nilearn/_utils/helpers.py:def _compare_version(version_a, operator, version_b):
nilearn/_utils/helpers.py:        error_msg = "'_compare_version' received an unexpected operator "
nilearn/plotting/displays/_slicers.py:from nilearn._utils import _compare_version, check_niimg_3d
nilearn/plotting/displays/_slicers.py:        if _compare_version(matplotlib.__version__, ">=", "1.6"):
nilearn/plotting/displays/_slicers.py:            if _compare_version(matplotlib.__version__, ">=", "1.6"):
nilearn/plotting/displays/_slicers.py:            if _compare_version(matplotlib.__version__, ">=", "1.6"):
nilearn/plotting/glass_brain.py:from nilearn._utils import _compare_version
nilearn/plotting/glass_brain.py:    if _compare_version(matplotlib.__version__, '>=', "2.0"):
nilearn/plotting/__init__.py:        from .._utils import _compare_version
nilearn/plotting/__init__.py:        if not _compare_version(
```

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #3793
- Closes #3706

### CODE REVIEW

**Comment by Remi-Gau** [comment]
Not sure I understand what to do about this one:

https://github.com/nilearn/nilearn/blob/5734dba905a275ec03dbcc39344a4e0a7e0e3b11/nilearn/plotting/displays/_slicers.py#L464-L472

File: nilearn/image/resampling.py [hunk 1]
```diff
@@ -300,8 +297,7 @@ def _resample_one_img(
 
     # Suppresses warnings in https://github.com/nilearn/nilearn/issues/1363
     with warnings.catch_warnings():
-        if _compare_version(scipy.__version__, ">=", "0.18"):
-            warnings.simplefilter("ignore", UserWarning)
+        warnings.simplefilter("ignore", UserWarning)
```
Comment (ymzayek) [inline nilearn/image/resampling.py:300]: Is it still relevant to suppress these warnings? Do they still occur in scipy>=1.6.0

File: nilearn/image/resampling.py [hunk 2]
```diff
@@ -316,8 +312,7 @@ def _resample_one_img(
     if has_not_finite:
         # Suppresses warnings in https://github.com/nilearn/nilearn/issues/1363
         with warnings.catch_warnings():
-            if _compare_version(scipy.__version__, ">=", "0.18"):
-                warnings.simplefilter("ignore", UserWarning)
+            warnings.simplefilter("ignore", UserWarning)
```
Comment (ymzayek) [inline nilearn/image/resampling.py:315]: same here

**Comment by ymzayek** [comment]
> Not sure I understand what to do about this one:
> 
> https://github.com/nilearn/nilearn/blob/5734dba905a275ec03dbcc39344a4e0a7e0e3b11/nilearn/plotting/displays/_slicers.py#L464-L472

It's not very clear from the comment because there has been a refactoring but from the git blame and the original issue https://github.com/matplotlib/matplotlib/issues/9280 I gathered that you can remove the second part of L465:
if data_2d is not None ~~and data_2d.min() is not np.ma.masked~~:

**Comment by ymzayek** [comment]
Can you also have a look at https://github.com/nilearn/nilearn/blob/5734dba905a275ec03dbcc39344a4e0a7e0e3b11/nilearn/surface/surface.py#L107-L108 and https://github.com/nilearn/nilearn/blob/5734dba905a275ec03dbcc39344a4e0a7e0e3b11/nilearn/surface/tests/test_surface.py#L573-L574
They both deal with np 1.15 even though the first one incorrectly references np 1.5

File: nilearn/image/resampling.py (hunk 1 above)
Comment (Remi-Gau) [reply nilearn/image/resampling.py:300]: let me check

Comment (Remi-Gau) [reply nilearn/image/resampling.py:300]: warnings are still there, so it does make sense to suppress the warning but I will add a comment because without the scipy reference in the if block if becomes unclear why we are suppressing it

Comment (Remi-Gau) [reply nilearn/image/resampling.py:300]: actually improved the filtering instead

**Comment by Remi-Gau** [comment]
> It's not very clear from the comment because there has been a refactoring but from the git blame and the original issue [matplotlib/matplotlib#9280](https://github.com/matplotlib/matplotlib/issues/9280) I gathered that you can remove the second part of L465: if data_2d is not None ~and data_2d.min() is not np.ma.masked~:

Thanks: let's try this then

File: nilearn/image/resampling.py [hunk 3]
```diff
@@ -300,8 +297,9 @@ def _resample_one_img(
 
     # Suppresses warnings in https://github.com/nilearn/nilearn/issues/1363
     with warnings.catch_warnings():
-        if _compare_version(scipy.__version__, ">=", "0.18"):
-            warnings.simplefilter("ignore", UserWarning)
+        warnings.filterwarnings(
+            "ignore", message=".*has changed in SciPy 0.18.*"
+        )
```
Comment (Remi-Gau) [inline nilearn/image/resampling.py:302]: figured that it'd be more specific rather than ignoring all user warnings, WDYT?

File: nilearn/surface/tests/test_surface.py [hunk 4]
```diff
@@ -562,8 +562,12 @@ def test_sample_locations_depth(depth, n_points):
...
     "depth,n_points",
-    [(None, 1), (None, 7), ([0.], 8), ([-1.], 8),
-     ([1.], 8), ([-1., 0., .5], 8)])
+    [(None, 1),
+     (None, 7),
+     ([0.], 8),
+     ([-1.], 8),
+     ([1.], 8),
+     ([-1., 0., .5], 8)])
 def test_sample_locations_between_surfaces(depth, n_points):
```
Comment (Remi-Gau) [inline nilearn/surface/tests/test_surface.py:571]: still need to change this test

**Comment by Remi-Gau** [comment]
> Can you also have a look at
> 
> https://github.com/nilearn/nilearn/blob/5734dba905a275ec03dbcc39344a4e0a7e0e3b11/nilearn/surface/surface.py#L107-L108
> and
> 
> https://github.com/nilearn/nilearn/blob/5734dba905a275ec03dbcc39344a4e0a7e0e3b11/nilearn/surface/tests/test_surface.py#L573-L574
> 
> They both deal with np 1.15 even though the first one incorrectly references np 1.5

could not figure out how to get those to work
must be missing something from the comment

**Review by bthirion — COMMENTED** [review]
LGTM so far.

**Comment by ymzayek** [comment]
@jeromedockes any pointers on these comments https://github.com/nilearn/nilearn/pull/3883#issuecomment-1666332525?

**Comment by jeromedockes** [comment]
> @jeromedockes any pointers on these comments [#3883 (comment)](https://github.com/nilearn/nilearn/pull/3883#issuecomment-1666332525)?

it works fine as it is (including with newer numpy versions) so we can just delete the comments

**Comment by Remi-Gau** [comment]
> > @jeromedockes any pointers on these comments [#3883 (comment)](https://github.com/nilearn/nilearn/pull/3883#issuecomment-1666332525)?
> 
> it works fine as it is (including with newer numpy versions) so we can just delete the comments

Good with me

**Comment by Remi-Gau** [comment]
> LGTM @Remi-Gau !
> 
> The codecov failure refers to just one line that's not covered (though not due to this PR), if it's easy to add a test let's do it here? WDYT?

added a dummy test to cover that piece of code: I love testing but not a big fan of adding tests just for coverage's.
In this it could be used mostly as unofficial doc to show dev how to use this function in testing.

I had look at the how deprecation is handled in general and it seems that we have quite a few deprecation helper functions that have gone mostly idle: so some refactoring seems to be needed

**Comment by ymzayek** [comment]
Fair enough, tbh I didn't take a close look at the function.

Yes I am for trying to standardize the way we handle deprecations

**Comment by ymzayek** [comment]
Oops just meant to comment not close this
